store/src/commonMain/kotlin/com/dropbox/external/store5/impl/ShareableMarket.kt

```diff
@@ -0,0 +1,431 @@
+            load<Output>(key)
+            refresh(key, fetcher)
+        }
+    }
+
+    private suspend fun <Output : Any> getOrSetBroadcast(key: Key): SomeBroadcast<Output> {
+
+        val result = scope.async {
+            if (broadcasts.access { it[key] == null }) {
+                broadcasts.access {
+                    it[key] = MutableSharedFlow(4)
+                    scope.launch {
+                        it[key]!!.emit(MarketResponse.Loading)
+                    }
+                }
+            }
+        }
+
+
+        result.await()
+        return requireBroadcast(key)
+    }
+
+    private fun <Output : Any> getBroadcast(key: Key): SomeBroadcast<Output>? =
+        broadcasts.access { it[key] } as? SomeBroadcast<Output>
+
+    private fun <Output : Any> requireBroadcast(key: Key): SomeBroadcast<Output> =
+        broadcasts.access { it[key]!! } as SomeBroadcast<Output>
+
+    private fun <Input : Any> requireWriteRequestQueue(key: Key): SomeWriteRequestQueue<Key, Input> =
+        writeRequests.access { it[key]!! } as SomeWriteRequestQueue<Key, Input>
+
+    private fun broadcasting(key: Key) = broadcasts.access { it[key] != null }
+
+    private fun addOrInitReadCompletions(key: Key, onCompletions: AnyReadCompletionsQueue) {
+        if (readCompletions.access { it[key] != null }) {
+            readCompletions.access { it[key]!!.addAll(onCompletions) }
+        } else {
+            readCompletions.access { it[key] = onCompletions }
+        }
+    }
+
+    private fun resetWriteRequestQueue(key: Key, queue: AnyWriteRequestQueue<Key>) {
+        writeRequests.access { it[key] = queue }
+    }
+
+    private fun <Input : Any, Output : Any> addOrInitWriteRequest(
+        key: Key,
+        updater: Updater<Key, Input, Output>
+    ) {
+        if (writeRequests.access { it[key] == null }) {
+            writeRequests.access { it[key] = ArrayDeque() }
+        }
+
+        writeRequests.access { it[key]!!.add(updater) }
+    }
+
+    private fun <Input : Any, Output : Any> getLatestWriteRequest(key: Key) =
+        writeRequests.access { it[key]?.last() } as Updater<Key, Input, Output>
+
+    private fun readInProgress(key: Key) = readCompletions.access { it[key].isNullOrEmpty().not() }
+
+    private suspend fun <Output : Any> lastStored(key: Key): Output? {
+        for (store in stores) {
+            try {
+                val last = (store.read(key) as Flow<Output?>).lastOrNull()
+                if (last != null) {
+                    return last
+                }
+            } catch (_throwable: Throwable) {
+                continue
+            }
+        }
+        return null
+    }
+
+    private suspend fun <Input : Any, Output : Any> eagerlyResolveConflicts(
+        key: Key,
+        request: PostRequest<Key, Input, Output>,
+        converter: Converter<Output, Input>
+    ): Boolean {
+
+        return try {
+            val lastStored = scope.async { lastStored<Output>(key) }
```

Comment:
We can get rid of this async. This is a private function so we control the coroutine scope of the caller, no need to redispatch.